### ISSUE TITLE
Possible leak with TLS

### ISSUE THREAD

**Issue by nbigaouette** [tokio-rs/tokio#2554]
**Version**

```txt
❯ ❯ rg tokio README.md
117:│   ├── tokio v0.2.21
141:│   │   └── tokio-macros v0.2.5
145:│   ├── tokio-tls v0.3.1
147:│   │   └── tokio v0.2.21 (*)
149:│   │   └── tokio v0.2.21 (*)
150:│   ├── tokio-util v0.3.1
156:│   │   └── tokio v0.2.21 (*)
158:│   │   └── tokio v0.2.21 (*)
168:└── tokio v0.2.21 (*)
```

(README.md is from https://github.com/nbigaouette/ldap3-issue47/blob/14381b9c8625e20426cb19e8a9f103fc/README.md)

**Platform**

```sh
❯ uname -a
Darwin nbmbp.local 19.4.0 Darwin Kernel Version 19.4.0: Wed Mar  4 22:28:40 PST 2020; root:xnu-6153.101.6~15/RELEASE_X86_64 x86_64
```

**Description**

When repeatadly establishing a TLS connection to an ldap server (using the `ldap3` crate), I have a ~55 kB leak on every connection.

I have a repository containing a minimal reproducible example. See the `tokio-report` branch there: https://github.com/nbigaouette/ldap3-issue47/tree/tokio-report

I used XCode's _Instruments_ to measure this (and the [`cargo-instruments`](https://github.com/cmyr/cargo-instruments) to drive the measurement). For what it's worth I've uploaded the Instruments trace file:
[ldap3-issue47_2020-05-21T121219.trace.zip](https://github.com/tokio-rs/tokio/files/4663547/ldap3-issue47_2020-05-21T121219.trace.zip)

The repo linked above also contains a screenshot and the first stack trace in the list. There isn't a single one though in this list but I'm not sure which one contains useful information.

I initially though it could be caused by the sync wrapper of `ldap3` over its async API. This wrapper will launch a tokio runtime, await the result and then exit (see https://docs.rs/ldap3/0.7.0-alpha.7/src/ldap3/sync.rs.html#43-57) so I've modified the example repo to use `#[tokio::main]` with the async API directly.

With this version of the example, I see the number of file descriptors increase in the Activity Monitor: 
<img width="671" alt="Screen Shot 2020-05-21 at 12 34 41 PM" src="https://user-images.githubusercontent.com/112988/82582085-809a0f80-9b5f-11ea-9133-5fdaa94303bc.png">
and the memory increase from 4.7 MB to 10 MB in 3 minutes. After 6 minutes the code fails with:
`Error: Io { source: Os { code: 24, kind: Other, message: "Too many open files" } }`.

Other posts I've made regarding this problem:
* https://users.rust-lang.org/t/memory-leak-when-continously-creating-tokio-runtimes-in-a-loop/42972
* https://github.com/inejge/ldap3/issues/47

**Comment by Darksonn** [comment]
To clear up potentially confusing things in the minimal example, I'd like to have the following changed:

1. Use [`delay_for`][1] instead of blocking the thread with `thread::sleep`.
2. Replace the [`drive!`][3] macro with its expansion.
```rust
tokio::spawn(async move {
    if let Err(e) = conn.drive().await {
        ldap::log::warn!("LDAP connection error: {}", e);
    }
});
```
Additionally, I would like to hear what happens if you `.await` the [`JoinHandle`][2] returned by the `tokio::spawn` that the macro expands to.

[1]: https://docs.rs/tokio/0.2/tokio/time/fn.delay_for.html
[2]: https://docs.rs/tokio/0.2/tokio/task/struct.JoinHandle.html
[3]: https://docs.rs/ldap3/0.7.0-alpha.7/ldap3/macro.drive.html

**Comment by nbigaouette** [comment]
I've updated the repo with your suggestions (https://github.com/nbigaouette/ldap3-issue47/tree/b5736dcd66d07e5248c73ecd5eb4d2469343d67b).

Commit b5736dcd66d07e5248c73ecd5eb4d2469343d67b does _not_ await the handle. With this I see the leak (see screenshot in the README at that commit). If I uncomment [the handle await](https://github.com/nbigaouette/ldap3-issue47/blob/b5736dcd66d07e5248c73ecd5eb4d2469343d67b/src/main.rs#L25) it deadlocks.

**Comment by Darksonn** [comment]
What if you `.await` the handle _after_ the `drop(ldap);` ?

**Comment by nbigaouette** [comment]
Same deadlock...

**Comment by Darksonn** [comment]
The documentation on [`ldap3::LdapConnAsync`][1] says "Dropping the last handle will automatically close the connection.", but that doesn't appear to happen here, then.

I'm guessing that is the issue.

[1]: https://docs.rs/ldap3/0.7.0-alpha.7/ldap3/struct.LdapConnAsync.html

**Comment by nbigaouette** [comment]
For now, I can reproduce the problem on macOS but only using TLS. My Windows machine is not ready yet, but I do see the increase in memory usage in production:

<img width="794" alt="Screen Shot 2020-05-21 at 8 50 09 PM" src="https://user-images.githubusercontent.com/112988/82702732-94697280-9c40-11ea-8650-36ec8fa28599.png">
<img width="785" alt="Screen Shot 2020-05-21 at 8 50 18 PM" src="https://user-images.githubusercontent.com/112988/82702733-95020900-9c40-11ea-99a3-105e8a600493.png">

From https://github.com/inejge/ldap3/issues/47:

> So far, it looks like an issue with Secure Transport bindings to `native-tls`. Maybe something in the way in which `ldap3` uses TLS or manages connections triggers the leak?

It definitely looks like something related to TLS.

> The documentation on [`ldap3::LdapConnAsync`](https://docs.rs/ldap3/0.7.0-alpha.7/ldap3/struct.LdapConnAsync.html) says "Dropping the last handle will automatically close the connection.", but that doesn't appear to happen here, then.
> 
> I'm guessing that is the issue.

This problem in `ldap3` should be fixed in its 2f88e03b9a7cd16b33a0a8feca9e472f, which is the version used for the above tests.

What would be a good tool to measure the memory usage (and ideally identify the leak) on Windows?
